Add waitany and waitall functions to wait multiple tasks at once

mrkn opened the pull request:
I would like to propose adding two functions, `waitany` and `waitall`, discussed in the issue #53226. These functions wait for multiple tasks at once. The `waitany` function blocks until one task finishes. The `waitall` function blocks until all tasks finish. There is an optional keyword argument, `failfast`, for the `waitall` function. The default of `failfast` is `false`. The `waitall` function will immediately stop if any task ends with an exception when the `failfast` is `true`.

This is my own implementation, but I have regrets about the type of the first argument. I wanted to represent a container type from which `Task` objects can be taken out using the `iterate` function, but it seems impossible in current Julia, so I used a union type of `AbstractVector`, `Tuple`, and `Set`. I would like to know if there is a better way to write this part.

mrkn committed:
Add waitany and waitall to wait multiple tasks

mrkn committed:
Reduce the number of working vectors

mrkn committed:
Rewrite with using Channel

mrkn committed:
Add test cases to Cover all argument types

mrkn commented on base/task.jl:
```diff
@@ -365,6 +365,60 @@ function wait(t::Task)
+waitany(tasks) = _wait_multiple(tasks)
+waitall(tasks; failfast=false) = _wait_multiple(tasks; all=true, failfast=failfast)
+
+function _wait_multiple(waiting_tasks::Union{AbstractVector{Task},AbstractSet{Task},Tuple{Task}}; all=false, failfast=false)::NTuple{2,Vector{Task}}
```
I guess `typeassert` can be used to remove the restriction of the type of the first argument like below, but I don't know whether this is a general programming pattern in Julia.  I cannot find any other use cases to check the `eltype` with `typeassert` in this repository.

```suggestion
function _wait_multiple(waiting_tasks; all=false, failfast=false)::NTuple{2,Vector{Task}}
    typeassert(eltype(waiting_tasks), Type{Task})
```

aviatesk replied:
Yeah, I believe it's best to avoid type annotations for arguments here. While using `eltype` is an option, we could also consider iterating over `waiting_tasks` and directly applying `typeassert` to each element.
```suggestion
function _wait_multiple(waiting_tasks; all=false, failfast=false)::NTuple{2,Vector{Task}}
    foreach(waiting_tasks) do @nospecialize t
        if !(t isa Task)
            error("Expected an iterator of `Task` object")
        end
    end
```
And then we could also enhance type stability in the main `enumerator` for-loop by allowing you to specify `t = t::Task`.
```julia
    for (i, t) in enumerate(waiting_tasks)
        t = t::Task
        ...
```

mrkn replied:
Thank you for your feedback.  I'd like to adopt your suggestion because it can accept inputs that consist only of tasks but have an `Any` eltype, such as `Vector{Any}`.  My version cannot accept such types of inputs.

aviatesk commented on base/task.jl:
```diff
@@ -365,6 +365,60 @@ function wait(t::Task)
+waitany(tasks) = _wait_multiple(tasks)
+waitall(tasks; failfast=false) = _wait_multiple(tasks; all=true, failfast=failfast)
+
+function _wait_multiple(waiting_tasks::Union{AbstractVector{Task},AbstractSet{Task},Tuple{Task}}; all=false, failfast=false)::NTuple{2,Vector{Task}}
```
Also, I wonder if the return type declaration is redundant, given that this function seems type stable with respect to the return values.

mrkn replied:
Yes, its my fault. I'll remove the return type.

mrkn commented on base/task.jl:
```diff
@@ -365,6 +365,60 @@ function wait(t::Task)
+waitany(tasks) = _wait_multiple(tasks)
+waitall(tasks; failfast=false) = _wait_multiple(tasks; all=true, failfast=failfast)
+
+function _wait_multiple(waiting_tasks::Union{AbstractVector{Task},AbstractSet{Task},Tuple{Task}}; all=false, failfast=false)::NTuple{2,Vector{Task}}
```
```suggestion
function _wait_multiple(waiting_tasks::Union{AbstractVector{Task},AbstractSet{Task},Tuple{Task}}; all=false, failfast=false)
```

mrkn committed:
Remove return type of _wait_multiple

mrkn commented on base/task.jl:
```diff
@@ -365,6 +365,60 @@ function wait(t::Task)
+waitany(tasks) = _wait_multiple(tasks)
+waitall(tasks; failfast=false) = _wait_multiple(tasks; all=true, failfast=failfast)
+
+function _wait_multiple(waiting_tasks::Union{AbstractVector{Task},AbstractSet{Task},Tuple{Task}}; all=false, failfast=false)
+    chan = Channel{Tuple{Int,Task}}(Inf)
```
```suggestion
function _wait_multiple(waiting_tasks; all=false, failfast=false)
    foreach(waiting_tasks) do @nospecialize t
        if !(t isa Task)
            error("Expected an iterator of `Task` object")
        end
    end

    chan = Channel{Tuple{Int,Task}}(Inf)
```

mrkn committed:
Remove 1st argument type of _wait_multiple

Instead of annotating the argument type, check the type of each actual element.

Co-authored-by: Shuhei Kadowaki <aviatesk@gmail.com>

mrkn commented on base/task.jl:
```diff
@@ -365,6 +365,66 @@ function wait(t::Task)
+    exception = false
+    nremaining::Int = 0
+
+    for (i, t) in enumerate(waiting_tasks)
+        push!(tasks, t)
```
```suggestion
    for (i, t) in enumerate(waiting_tasks)
        t = t::Task
        push!(tasks, t)
```

mrkn committed:
Specify type of element that comes from iteration for type stability

Co-authored-by: Shuhei Kadowaki <aviatesk@gmail.com>

vtjnash commented on base/task.jl:
```diff
@@ -365,6 +365,67 @@ function wait(t::Task)
+waitall(tasks; failfast=false) = _wait_multiple(tasks; all=true, failfast=failfast)
+
+function _wait_multiple(waiting_tasks; all=false, failfast=false)
+    foreach(waiting_tasks) do @nospecialize t
+        if !(t isa Task)
+            error("Expected an iterator of `Task` object")
+        end
+    end
+
```
Iterating `waiting_tasks` may change the contents (e.g. consuming it) or they may change afterwards, so this isn't sufficient or helpful to check upfront
```suggestion
```

aviatesk replied:
Indeed, that would be the case in worst-case scenarios. The intention of this code seems to be to avoid scheduling new `Task` when given invalid `waiting_tasks`. So I'd propose adjusting the main `enumerate` iteration as follows?:
```julia
function _wait_multiple(waiting_tasks; all=false, failfast=false)
    tasks = Task[]
    done_mask = Bool[]
    exception = false
    nremaining::Int = 0

    for (i, t) in enumerate(waiting_tasks)
        t isa Task || error("Expected an iterator of `Task` object")
        push!(tasks, t)
        if istaskdone(t)
            push!(done_mask, true)
            exception |= istaskfailed(t)
        else
            push!(done_mask, false)
            nremaining += 1
        end
    end

    chan = Channel{Tuple{Int,Task}}(Inf)
    for (i, done) in enumerate(done_mask)
        if !done
            t = tasks[i]
            schedule(Task(() -> begin
                _wait(t)
                put!(chan, (i, t))
            end))
        end
    end

    ...
```

mrkn replied:
@aviatesk Thank you for suggesting the code.  This can apply to a consuming task stream and avoid invalid input handling. I'll apply this.

mrkn replied:
I've applied this at 1bd4973.

vtjnash commented on base/task.jl:
```diff
@@ -365,6 +365,67 @@ function wait(t::Task)
+        else
+            nremaining += 1
+            push!(done_mask, false)
+            schedule(Task(() -> begin
+                _wait(t)
+                put!(chan, (i, t))
+            end))
```
performance nit: this is a more optimized internal version of the same:
```suggestion
            t2 = @task put!(chan, (i, t))
            t2.sticky = false
            _wait2(t, t2)
```

mrkn replied:
The commit 1bd4973 includes this suggestion.

vtjnash commented on base/task.jl:
```diff
@@ -365,6 +365,67 @@ function wait(t::Task)
+    if nremaining == 0
+        close(chan)
+        return tasks, Task[]
+    elseif any(done_mask) && (!all || (failfast && exception))
+        close(chan)
+        return tasks[done_mask], tasks[.~done_mask]
```
It looks like the `any` option can leak a non-trivial amount of resources on each call. There is not a particularly good API for avoiding that either. It needs to go back through all of the `_wait2` tasks and delete them from the donenotify list if they are present there, using something like:
```julia
@lock t.donenotify Base.list_deletefirst!((t.donenotify::ThreadSynchronizer).waitq, t2)
```

mrkn replied:
@vtjnash Thank you for your advice. I'll insert the cleanup code.

mrkn committed:
Support inputs that can be iterated only once

Co-authored-by: Shuhei Kadowaki <aviatesk@gmail.com>
Co-authored-by: Jameson Nash <vtjnash@gmail.com>

mrkn committed:
Delete waiters from waitq of each remaining task

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

mrkn commented on base/task.jl:
```diff
@@ -420,7 +423,12 @@ function _wait_multiple(waiting_tasks; all=false, failfast=false)
         return tasks, Task[]
     else
-        return tasks[done_mask], tasks[.~done_mask]
+        remaining_tasks = tasks[.~done_mask]
+        for t in remaining_tasks
+            waiter = pop!(waiter_tasks, t)
+            @lock t.donenotify Base.list_deletefirst!((t.donenotify::ThreadSynchronizer).waitq, waiter)
+        end
+        return tasks[done_mask], remaining_tasks
```
@vtjnash I've added this clean-up code to delete waiter tasks from the `waitq` of each remaining task.
I introduced the `waiter_tasks` dictionary to obtain the corresponding waiter task `waiter` from the target task `t`.

vtjnash reviewed (commented):
This seems quite great. It has got lots of annoying details if anyone was trying to emulate it themself, but seems fairly straightforward internally once you work out those details. I think we are ready to merge it, unless there is anything else of concern with it from anyone? One observation I might suggest is that we add another flag `throw` (related to failfast) that defines whether it should fail with `CompositeException` when `exception` is set (independent of failfast)?

This may also be a good drop-in replacement for the `Experiment.sync_end` implementation, with waitall and failfast, which may be a good demo of it in use?

vtjnash commented on base/task.jl:
```diff
@@ -365,6 +365,73 @@ function wait(t::Task)
+    waiter_tasks = Dict{Task, Task}()
+
+    for (i, done) in enumerate(done_mask)
+        if !done
```
performance nit: add a check here to break if something finishes early?
```suggestion
        done && continue
        if istaskdone(t)
            t = tasks[i]
            exception |= istaskfailed(t)
            done_mask[i] = true
            exception && failfast && break
        else
```

vtjnash commented on base/task.jl:
```diff
@@ -365,6 +365,73 @@ function wait(t::Task)
+    for (i, t) in enumerate(waiting_tasks)
+        t isa Task || error("Expected an iterator of `Task` object")
+        push!(tasks, t)
+        if istaskdone(t)
+            push!(done_mask, true)
+            exception |= istaskfailed(t)
+        else
+            push!(done_mask, false)
```
performance nit: enumerating `waiting_tasks` might take a long time (e.g. if it is a Channel). It could thus help being more efficient in the next stage to split off the check for done-ness until that finishes being enumerated:
```suggestion
    end
    
    done_mask = falses(length(tasks))
    for (i, t) in enumerate(tasks)
        if istaskdone(t)
            done_mask[i] = true
            exception |= istaskfailed(t)
        else
            done_mask[i] = false
```

vtjnash commented on base/task.jl:
```diff
@@ -365,6 +365,73 @@ function wait(t::Task)
+
+function _wait_multiple(waiting_tasks; all=false, failfast=false)
+    tasks = Task[]
+    done_mask = Bool[]
+    exception = false
+    nremaining::Int = 0
```
```suggestion
```

vtjnash commented on base/task.jl:
```diff
@@ -365,6 +365,73 @@ function wait(t::Task)
+    for (i, t) in enumerate(waiting_tasks)
+        t isa Task || error("Expected an iterator of `Task` object")
+        push!(tasks, t)
+        if istaskdone(t)
+            push!(done_mask, true)
+            exception |= istaskfailed(t)
+        else
+            push!(done_mask, false)
+            nremaining += 1
```
performance nit: enumerating `waiting_tasks` might take a long time (e.g. if it is a Channel). It could thus help being more efficient in the next stage to split off the check for done-ness until that finishes being enumerated:
```suggestion
    end

    exception = false
    nremaining::Int = length(tasks)
    done_mask = falses(nremaining)
    for (i, t) in enumerate(tasks)
        if istaskdone(t)
            done_mask[i] = true
            exception |= istaskfailed(t)
            nremaining -= 1
        else
            done_mask[i] = false
```

vtjnash commented on base/task.jl:
```diff
@@ -365,6 +365,73 @@ function wait(t::Task)
+        return tasks[done_mask], tasks[.~done_mask]
+    end
+
+    chan = Channel{Tuple{Int,Task}}(Inf)
+    waiter_tasks = Dict{Task, Task}()
```
Do you think it be sensible to use the index for this, for a vector rather than a map?
```suggestion
    chan = Channel{Int}(Inf)
    waiter_tasks = Vector{Task}(undef, length(tasks))
    # or perhaps, `fill(current_task(), length(tasks))`, and use `current_task()` as a sentinel value instead of using `_unsetindex!`?
```
later then instead of
```
        i, t = take!(chan)
```
using the `tasks` vector
```
        i = take!(chan)
        t = tasks[i]
```
and later instead of
```
for t in remaining_tasks
```
use
```
for t in waiter_tasks[.~done_mask]
end
```

mrkn replied:
I used `current_task()` as a sentinel.

mrkn committed:
Fix for performance

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

mrkn committed:
Split type checking and examining loops

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

vtjnash commented on base/task.jl:
```diff
@@ -365,6 +365,77 @@ function wait(t::Task)
+        exception |= istaskfailed(t)
+        nremaining -= 1
+
+        if nremaining == 0 || (!all || failfast && exception)
```
another small possible improvement here is to try to empty the channel before breaking, if it can be done non-blocking:
```suggestion
        # stop early if requested, unless there is something already immediately already ready to consume from the channel (using a race-y check)
        if (!all || (failfast && exception)) && !isready(chan)
```

vtjnash commented on base/task.jl:
```diff
@@ -365,6 +365,77 @@ function wait(t::Task)
+        end
+    end
+
+    while true
```
```suggestion
    while nremaining > 0
```

vtjnash commented on base/task.jl:
```diff
@@ -365,6 +365,77 @@ function wait(t::Task)
 end
 
+# Wait multiple tasks
+waitany(tasks) = _wait_multiple(tasks)
+waitall(tasks; failfast=false) = _wait_multiple(tasks; all=true, failfast=failfast)
```
Can you add docstrings to these, add them to docs/src/parallel.md, and add a mention of this feature in NEWS.md? I am quite happy with where this PR is, so just these few remaining details to cleanup. Thanks for pursuing this!

mrkn replied:
Yes, I intended to add docstring to these after finishing the fixes to the implementation.

I think I can finish the remaining fixes and writing docs in this week.  Thank you for your valuable feedback! Without them, I couldn’t  complete this pull request.

mrkn committed:
Stop using needless enumerate

mrkn committed:
Optimize for waitall with failfast=false

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

mrkn commented on base/task.jl:
```diff
@@ -365,6 +365,86 @@ function wait(t::Task)
+        push!(tasks, t)
+    end
+
+    if all && !failfast
+        # Force everything to finish synchronously for the case of waitall
+        # with failfast=false
+        for t in tasks
+            _wait(t)
+        end
+        return tasks, Task[]
+    end
```
@vtjnash I've added this loop as the answer to your comment https://github.com/JuliaLang/julia/pull/53341#discussion_r1509354697.
I think this is better than inserting a fully synchronous path in the loop below.

mrkn committed:
Use vector for managing waiters

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

mrkn committed:
Add channel emptiness check

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

mrkn committed:
Insert done check in waiter creation loop

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

mrkn committed:
Stop using kwargs in _wait_multiple

mrkn committed:
Add throw keyword argument in waitall

mrkn committed:
Add throw keyword argument in waitany

quinnj commented:
I've been excited seeing this PR! Just to clarify (I know docs are coming, but until then); does `waitany` return the Task that finished first? I ask because there have been a few scenarios where I've wanted a golang-style `select` and have come up with subpar versions of `waitany`. Just curious!

mrkn committed:
Add docstrings of waitany and waitall

mrkn commented:
@quinnj `waitany` is more similar to `select` system call than golang's `select`.  `waitany` may return more than one completed tasks in the first return vector instead of only one completed task.  But, you can easily emulate golang's `select` by using `waitany`, I think.

mrkn commented:
@vtjnash I've finished supporting `throw` option and writing docstrings.  Could you please check the last three commits?

vtjnash commented:
A `select` in go waits on (multiple) channels (which this cannot do, as this doesn't fetch and return values), while this waits on Tasks (which go is not capable of doing)

vtjnash commented on base/task.jl:
```diff
@@ -378,12 +406,19 @@ function _wait_multiple(waiting_tasks, all=false, failfast=false)
+            exception |= istaskfailed(t)
+        end
+        if exception && throwexc
+            exceptions = [t.exception for t in tasks if istaskfailed(t)]
```
```suggestion
            exceptions = [TaskFailedException(t) for t in tasks if istaskfailed(t)]
```

vtjnash commented on base/task.jl:
```diff
@@ -378,12 +406,19 @@ function _wait_multiple(waiting_tasks, all=false, failfast=false)
     end
 
     if all && !failfast
```
```suggestion
    if (all && !failfast) || length(t) <= 1
```

vtjnash commented on base/task.jl:
```diff
@@ -451,7 +491,13 @@ function _wait_multiple(waiting_tasks, all=false, failfast=false)
-        return tasks[done_mask], tasks[remaining_mask]
+        done_tasks = tasks[done_mask]
+        if throwexc && exception
+            exceptions = [t.exception for t in done_tasks if istaskfailed(t)]
```
```suggestion
            exceptions = [TaskFailedException(t) for t in done_tasks if istaskfailed(t)]
```

vtjnash commented on test/threads_exec.jl:
```diff
@@ -1210,19 +1210,36 @@ end
 
+    function teardown(tasks, event)
+        notify(event)
+        yield()
```
This does not seem relevant
```suggestion
```

vtjnash commented on test/threads_exec.jl:
```diff
@@ -1210,19 +1210,36 @@ end
+    function teardown(tasks, event)
+        notify(event)
+        yield()
+        wait(tasks[3])
```
```suggestion
        waitall(tasks, throw=true)
```

vtjnash commented on test/threads_exec.jl:
```diff
@@ -1257,9 +1274,36 @@ end
+                tasks, event = create_tasks()
+                push!(tasks, Threads.@spawn error("Error"))
+
+                sleep(0.1)
```
```suggestion
                wait(t[1])
                wait(t[2])
```

vtjnash commented on base/task.jl:
```diff
@@ -366,10 +366,38 @@ function wait(t::Task)
+
+Wait until at least one of the given tasks have been completed.
+
+If `throw` is `true`, the `CompositeException` will be thrown when one of the
+completed tasks completes with an exception.
+
+The return value consists of two task vectors. The first one consists of
```
wording nit: use a consistent declarative tense here. for example:
```suggestion
If `throw` is `true`, throw `CompositeException` when one of the
completed tasks completes with an exception.

Return two task vectors: the first one containing the completed tasks,
the second containing uncompleted tasks.
```

mrkn replied:
I've fixed the tense in docstrings of waitany and waitall.

mrkn committed:
Wait single task synchronously

mrkn committed:
Use TaskFailedException

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

mrkn committed:
Stop using sleep in test

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

mrkn committed:
Remove needless yield call

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

mrkn committed:
Wait all three tasks in teardown function in test

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

mrkn committed:
Use consistent declarative tense in docstring

mrkn committed:
Add waitany and waitall in doc/src/base/parallel.md

mrkn committed:
Add waitany and waitall in NEWS.md

mrkn commented on base/task.jl:
```diff
@@ -365,6 +365,142 @@ function wait(t::Task)
+# Wait multiple tasks
+
+"""
+    waitany(tasks; throw=false) -> (done_tasks, remaining_tasks)
```
I've done. Also, I've added the entry in NEWS.md.

mrkn commented on NEWS.md:
```diff
@@ -37,6 +37,7 @@ New library functions
 
 * `logrange(start, stop; length)` makes a range of constant ratio, instead of constant step ([#39071])
 * The new `isfull(c::Channel)` function can be used to check if `put!(c, some_value)` will block. ([#53159])
+* `waitany(tasks; throw=false)` and `waitall(tasks; failfast=false, throw=false)` which wait multiple tasks at once ([#53341])
```
@vtjnash Is this the right place for these functions?

vtjnash replied:
Yes, this looks like a reasonable spot for it. Style nit: they usually end in a `.`, though someone goes through and does a formatting cleanup anyways during the release, so it isn't vital to get these exactly right

mrkn replied:
I see. I'll fix this.

vtjnash commented on base/task.jl:
```diff
@@ -365,6 +365,142 @@ function wait(t::Task)
+
+The return value consists of two task vectors. The first one consists of
+completed tasks, and the other consists of uncompleted tasks.
+"""
```
```suggestion

!!! warning
    This may scale poorly compared to writing code that uses multiple individual tasks that
    each runs serially, since this needs to scan the list of `tasks` each time and
    synchronize with each one every time this is called. Or consider using
    [`waitall(tasks; failfast=true)`](@ref waitall) instead.
"""
```

mrkn replied:
I also think this should be a valuable note for users.

vtjnash replied:
It looks like you added this, then force-pushed it away?

mrkn replied:
Oops!

vtjnash commented on base/task.jl:
```diff
@@ -365,6 +365,142 @@ function wait(t::Task)
+waitany(tasks; throw=false) = _wait_multiple(tasks, throw)
+
+"""
+    waitall(tasks; failfast=false, throw=false) -> (done_tasks, remaining_tasks)
```
I wonder if we should switch these defaults? `throw` is the default behavior for `wait`, so I think that would be more consistent with it. It seems like `failfast` is often also a commonly desired behavior, though I am less certain of whether that is a good choice of default. (we can also merge and then change our minds about the default anytime before the next release)
```suggestion
    waitall(tasks; failfast=true, throw=true) -> (done_tasks, remaining_tasks)
```

mrkn replied:
I agree with you about `throw`'s default value.  The consistency with the existing APIs is critical.

Speaking of `failfast`, I don't know a strong preference for it.  I just selected `false` based on my feelings.  Simple `waitall(tasks)` seems to be waiting for all the tasks to be completed for me.  But, the default value should be selected for usefulness.  If it seems that `failfast=true` would occur more frequently, I'd like to change the default value to `true`.

mrkn replied:
I've added a commit to change the default values of `throw` and `failfast`.

mrkn commented on base/task.jl:
```diff
@@ -365,6 +365,142 @@ function wait(t::Task)
+"""
+waitall(tasks; failfast=false, throw=false) = _wait_multiple(tasks, throw, true, failfast)
+
+function _wait_multiple(waiting_tasks, throwexc=false, all=false, failfast=false)
+    tasks = Task[]
+
+    for t in waiting_tasks
+        t isa Task || error("Expected an iterator of `Task` object")
+        push!(tasks, t)
+    end
+
```
@vtjnash How do you think split `_wait_multiple` for `AbstractVector{Task}` and other types?

```suggestion
function _wait_multiple(waiting_tasks, throwexc=false, all=false, failfast=false)
    tasks = Task[]

    for t in waiting_tasks
        t isa Task || error("Expected an iterator of `Task` object")
        push!(tasks, t)
    end
    
    return _wait_multiple(tasks, throwexc, all, failfast)
end

function _wait_multiple(tasks::AbstractVector{Task}, throwexc=false, all=false, failfast=false)
```

vtjnash replied:
To avoid the copy? That makes sense to me. Though I don't know if all AbstractVector support the bitmasking operations, I think that should be valid.

As food for thought also, in a later PR, I am thinking we should also seek to extend this to allow any other waitable events too, not just Tasks (e.g. Channels, Event, AsyncEvent, Timer, etc.) and either add the `throwexc` kwarg to those other `wait` functions (instead of calling it `_wait`) or rename `_wait` to something like `waitready`

mrkn replied:
> Though I don't know if all AbstractVector support the bitmasking operations, I think that should be valid.

I guess `a[mask]` is processed by `_getindex(IndexStyle(a), a, LogicalIndex(mask)...)`, which then reaches `_unsafe_getindex` defined in multidimensional.jl, so all AbstractVector support the bitmasking operations.

> As food for thought also, in a later PR, I am thinking we should also seek to extend this to allow any other waitable events too, not just Tasks

I completely agree. As I wrote in issue #53226, I also thought that supporting waitable objects other than Tasks might be a good idea.

> either add the `throwexc` kwarg to those other `wait` functions (instead of calling it `_wait`) or rename `_wait` to something like `waitready`

I think the non-throw version might be used as frequently as `wait` in real-world application development.

mrkn replied:
I submitted a new PR #53685 for introducing `throw` option in `wait(::Task)` function.

mrkn committed:
Change default argument values

mrkn committed:
Add usage note of waitall in docstring

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

quinnj commented:
Thanks for all the persistence and effort here @mrkn! Really great to get this in!

IanButterworth commented:
Adding backport to 1.11 because we want this in 1.11 for https://github.com/JuliaLang/julia/pull/54674

IanButterworth commented:
https://github.com/JuliaLang/julia/pull/54760#pullrequestreview-2109082540 means we no longer need `waitany` on 1.11 so I'm removing the backport label
